### ISSUE TITLE
64 bit vertex support

### ISSUE THREAD

**Issue by nasamaher** [hageldave/JPlotter#25]
Hi, 

First, I'm very impressed with this codebase - thank you for your contribution!

I have successfully prototyped 64 bit GLSL support for line vertices in a forked JPlotter in order to support some high density data sets we're working with and would love to have it added to JPlotter.

Even thought the current _SegmentDetails_ accepts doubles, they currently get downcast to floats for the GPU and this doesn't have enough precision when zooming in.  I understand there might be workaround tricks like using doubles for your transform matrix calculation and rescaling my coordinates, but this seemed like a nice general solution.  On my GPU I didn't see a performance difference.

Overall it was only a handful of line changes but it requires alternate shaders and requires OpenGL 4.3 (but there could be a graceful degradation back to floats if not available)

I'm an old timer (pre-GLSL) OpenGLer and had to do some learning - this may be obvious to implement - but if you want to see my notes I'd be happy to share.  I only did a non-elegant prototype for Lines.

Regards,

![32](https://user-images.githubusercontent.com/1706941/118853155-72312180-b8a1-11eb-9911-1d87a3fcaff5.png)

![64](https://user-images.githubusercontent.com/1706941/118853165-752c1200-b8a1-11eb-9de8-4203d0272a87.png)

Steve

**Comment by nasamaher** [comment]
I should comment that our GPU is quite beefy (P6000) and I'm not implying there won't be a performance hit on other GPUs.  I would think a user would naturally select between the implementations by using floats vs doubles when creating lines.

**Comment by hageldave** [comment]
Hey there,

Thanks for posting here! I've come across this issue of too little precision before and as you mentioned, there are a few possible work arounds which would solve this, but require quite some implementation effort (e.g. rescaling coordinates) to work with the coordinate system. 
I'm definitely curious about the GL double precision variant of the `Lines` and `LinesRenderer` classes! Maybe the prototypical adaption carry over easily to the other renderers as well.

How come such a high version of GL is required? I thought double precision was supported starting from 4.0.

**Comment by nasamaher** [comment]
Hi hageldave,

You are right about OpenGL version:  the API calls work with GL40., although the GLSH scripts require #version 410.  I have implemented `Lines `and `Points`, with a crude binary switch (from an external class) to choose float vs double.  I've attached the relevant files.  I'm sure you'll find a nice way to auto-trigger the right shader and code - e.g., use the vertex types the user provides.  Let me know if you have any questions.

![image](https://user-images.githubusercontent.com/1706941/119033956-d9230900-b97b-11eb-8da0-d4476d997f10.png)


[64bit.zip](https://github.com/hageldave/JPlotter/files/6518220/64bit.zip)

**Comment by hageldave** [comment]
@nasamaher thanks for contributing your code :) I will make the other GL renderers (Text, Triangle, Curves) also capable of using double precision and come up with a mechanism for choosing/autosetting the precision.

I found a bug that you may want to fix in your own code base. It has to do with dashed line rendering. I fixed it already in the new branch, see https://github.com/hageldave/JPlotter/blob/164d7db08bad156588501bc509ee575585242e5c/jplotter/src/main/java/hageldave/jplotter/renderables/Lines.java#L633 
here the parameters `scaleX` and `scaleY` need to be passed to the methods instead of passing 1,1

**Comment by nasamaher** [comment]
Happy to contribute.  Thanks for the bug warning.

**Comment by hageldave** [comment]
Double precision in GL rendering has been implemented for all of the 'basic' renderers (lines, points, curves, triangles, text).
The respective renderers, as well as `CompleteRenderer` and `ChainedRenderer`, provide a method `setGLDoublePrecisionEnabled(boolean enable)` to enable double precision rendering. We chose this mechanic over the alternative of using a flag in the `Renderable` (i.e. specifying in a `Lines` object that it wants to be rendered with double precision) since it would cause costly changes of render programs while running the render routine of a `Renderer`

This led to a few breaking changes in the API of `GenericRenderer` and its subclasses (i.e. the basic renderers) and corresponding `Renderable` implementations (i.e. `Lines`, `Curves`, `Points`, ...) related to the render and updateGL calls. These are internal however, and most likely nothing a user of this library would touch anyway.

This functionality is now on the master branch as of 8480130c4fdae5c1f77bb44fced8e9dcc14bbae0 and will be part of the next release of JPlotter.

There is also a little test class for checking out the effects of double precision rendering. Zoom in until it gets messy, then turn on double precision in the GUI. https://github.com/hageldave/JPlotter/blob/8480130c4fdae5c1f77bb44fced8e9dcc14bbae0/jplotter/src/test/java/hageldave/jplotter/PrecisionTest.java#L24

@nasamaher I hope you are happy with this solution ;)
